include/seqan/journaled_string_tree/jst_extension_base.h

```diff
@@ -0,0 +1,313 @@
+inline typename GetPatternState<TExtension>::Type &
+state(JstExtensionBase<TExtension, TCxtPosition> & extBase)
+{
+    return extBase._state;
+}
+
+template <typename TExtension, typename TCxtPosition>
+inline typename GetPatternState<TExtension const>::Type &
+state(JstExtensionBase<TExtension, TCxtPosition> const & extBase)
+{
+    return extBase._state;
+}
+
+// ----------------------------------------------------------------------------
+// Function setState();
+// ----------------------------------------------------------------------------
+
+template <typename TExtension, typename TCxtPosition>
+inline void
+setState(JstExtensionBase<TExtension, TCxtPosition> & extBase,
+         typename GetPatternState<TExtension>::Type && state)
+{
+    extBase._state = state;
+}
+
+// ----------------------------------------------------------------------------
+// Function getObservedValue();
+// ----------------------------------------------------------------------------
+
+template <typename TExtension, typename TCxtPosition>
+inline auto
+getObservedValue(JstExtensionBase<TExtension, TCxtPosition> & extBase) -> decltype(state(extBase))
+{
+    return state(extBase);
+}
+
+template <typename TExtension, typename TCxtPosition>
+inline auto
+getObservedValue(JstExtensionBase<TExtension, TCxtPosition> const & extBase) -> decltype(state(extBase))
+{
+    return state(extBase);
+}
+
+// ----------------------------------------------------------------------------
+// Function setObservedValue();
+// ----------------------------------------------------------------------------
+
+template <typename TExtension, typename TCxtPosition, typename TValue>
+inline void
+setObservedValue(JstExtensionBase<TExtension, TCxtPosition> & extBase,
+                 TValue && val)
+{
+    setState(extBase, std::forward<TValue>(val));
+}
+
+// ----------------------------------------------------------------------------
+// Function run()
+// ----------------------------------------------------------------------------
+
+template <typename TExtension, typename TCxtPosition,
+          typename TTraverser,
+          typename TDelegate>
+inline auto
+run(JstExtensionBase<TExtension, TCxtPosition> & extension,
+    TTraverser const & traverser,
+    TDelegate && delegate) -> decltype(impl::run(extension._derived, traverser, TCxtPosition()).first)
+{
+    auto res = impl::run(extension._derived, traverser, TCxtPosition());
+    if (res.second)
+        delegate();
+    return res.first;
+}
+
+// ----------------------------------------------------------------------------
+// Function find()
+// ----------------------------------------------------------------------------
+
+template <typename TContainer, typename TSpec,
+          typename TAlgorithm,
+          typename TDelegate,
+          typename TObserver>
+inline void
+_find(TraverserImpl<TContainer, JstTraversalSpec<TSpec> > & traverser,
+      TAlgorithm & algorithm,
```

Comment:
I would propose to use either _ or namespace impl, but not both for consistency. What do you think?